core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java

```diff
@@ -79,31 +81,16 @@ void setChunkSize(ByteSizeValue chunkSize) { // only settable for tests
     }
 
     public void resync(IndexShard indexShard, ActionListener<ResyncTask> listener) {
-        final Translog.View view = indexShard.acquireTranslogView();
-        ActionListener<ResyncTask> wrappedListener = new ActionListener<ResyncTask>() {
-            @Override
-            public void onResponse(ResyncTask resyncTask) {
-                try {
-                    view.close();
-                } catch (IOException e) {
-                    onFailure(e);
-                }
-                listener.onResponse(resyncTask);
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                try {
-                    view.close();
-                } catch (IOException inner) {
-                    e.addSuppressed(inner);
-                }
-                listener.onFailure(e);
-            }
-        };
         try {
             final long startingSeqNo = indexShard.getGlobalCheckpoint() + 1;
-            Translog.Snapshot snapshot = view.snapshot(startingSeqNo);
+            Translog.Snapshot snapshot = indexShard.getTranslog().newSnapshotFromMinSeqNo(startingSeqNo);
+            listener = chain(wrap(r -> snapshot.close(), e -> {
```

Comment:
These things are subjecting. I personally prefer the wrapping because it allows to ignore the crud of wrapping and focus on the functionality. Same goes for not re-assigning the listener (I prefer my version as it the complexities are dealt with in the same place, when we wrap). I don't feel strongly about it and will happily go along with your version.